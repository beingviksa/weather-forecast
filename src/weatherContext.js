import { createContext, useContext, useState } from "react";

export const WeatherContext = createContext();

export const WeatherContextProvider = ({ children }) => {
  const [searchResults, setSearchResults] = useState([]);
  const [consolidatedData, setConsolidatedData] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState("");
  return (
    <WeatherContext.Provider
      value={{
        searchResults,
        setSearchResults,
        consolidatedData,
        setConsolidatedData,
        isLoading,
        setIsLoading,
        error,
        setError,
      }}>
      {children}
    </WeatherContext.Provider>
  );
};

export const useWeatherContext = () => {
  return useContext(WeatherContext);
};
