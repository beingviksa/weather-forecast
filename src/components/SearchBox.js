import { useState } from "react";
import { useNavigate, Link } from "react-router-dom";
import cities from "../lib/city.list.json";
import { useWeatherContext } from "../weatherContext";

const SearchBox = ({ showBackButton = false }) => {
  const navigate = useNavigate();
  const [query, setQuery] = useState("");
  const { searchResults, setSearchResults } = useWeatherContext();

  const onChange = (e) => {
    const { value } = e.target;
    setQuery(value);

    let matchingCities = [];

    if (value.length > 3) {
      for (let city of cities) {
        if (matchingCities.length >= 5) {
          break;
        }
        const match = city.name.toLowerCase().startsWith(value.toLowerCase());
        if (match) {
          const cityData = {
            ...city,
            slug: `${city.name.toLowerCase().replace(/ /g, "-")}-${city.id}`,
          };
          matchingCities.push(cityData);
        }
      }
    }
    setSearchResults(matchingCities);
  };

  const handleClick = (city) => {
    setQuery("");
    navigate(`/location/${city.slug}`);
  };

  return (
    <div className="search">
      {showBackButton && (
        <Link to="/" className="back-link">
          ‚Üê Home
        </Link>
      )}

      <input
        type="text"
        value={query}
        onChange={onChange}
        placeholder="Seach for a city..."
        data-testid="search-box"
      />

      {query.length > 3 && (
        <ul>
          {searchResults.length > 0 ? (
            searchResults.map((city, index) => (
              <li key={index}>
                <a onClick={() => handleClick(city)}>
                  {city.name}
                  {city.state ? `, ${city.state}` : ""}
                  <span>({city.country})</span>
                </a>
              </li>
            ))
          ) : (
            <li className="search__no-results">No results</li>
          )}
        </ul>
      )}
    </div>
  );
};

export default SearchBox;
