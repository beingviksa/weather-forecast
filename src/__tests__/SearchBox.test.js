import { render, act, fireEvent } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import SearchBox from "../components/SearchBox";
import { WeatherContext } from "../weatherContext";

describe("<SearchBox/>", () => {
  const props = {
    showBackButton: false,
  };

  const contextData = {
    searchResults: [
      [
        {
          id: 2643734,
          name: "Londonderry County Borough",
          state: "",
          country: "GB",
          coord: {
            lon: -7.30917,
            lat: 54.997211,
          },
          slug: "londonderry-county-borough-2643734",
        },
        {
          id: 2643743,
          name: "London",
          state: "",
          country: "GB",
          coord: {
            lon: -0.12574,
            lat: 51.50853,
          },
          slug: "london-2643743",
        },
        {
          id: 4030939,
          name: "London Village",
          state: "",
          country: "KI",
          coord: {
            lon: -157.475021,
            lat: 1.98487,
          },
          slug: "london-village-4030939",
        },
        {
          id: 4119617,
          name: "London",
          state: "AR",
          country: "US",
          coord: {
            lon: -93.25296,
            lat: 35.328972,
          },
          slug: "london-4119617",
        },
        {
          id: 4298960,
          name: "London",
          state: "KY",
          country: "US",
          coord: {
            lon: -84.08326,
            lat: 37.128979,
          },
          slug: "london-4298960",
        },
      ],
    ],
    setSearchResults: jest.fn(),
  };
  it("render the SearchBox Component", async () => {
    const { container } = render(
      <WeatherContext.Provider value={contextData}>
        <Router>
          <SearchBox {...props} />
        </Router>
      </WeatherContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should set the results on change event", async () => {
    const { getByTestId } = render(
      <WeatherContext.Provider value={contextData}>
        <Router>
          <SearchBox {...props} />
        </Router>
      </WeatherContext.Provider>
    );

    const searchText = getByTestId("search-box");
    act(() => {
      fireEvent.change(searchText, { target: { value: "London" } });
    });

    expect(contextData.setSearchResults).toBeCalled();
  });
});
