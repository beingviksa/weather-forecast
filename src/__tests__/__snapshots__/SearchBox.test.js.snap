// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox/> render the SearchBox Component 1`] = `
<div>
  <div
    class="search"
  >
    <input
      data-testid="search-box"
      placeholder="Seach for a city..."
      type="text"
      value=""
    />
  </div>
</div>
`;
